src/Repository/DefaultUnleashRepository.php

```diff
@@ -239,19 +247,161 @@ private function parseFeatures(array $body): array
+                    $dependencies = [];
+                    foreach ($unresolvedItem->getDependencies() as $dependency) {
+                        if (!$dependency instanceof UnresolvedFeatureDependency) {
+                            continue;
+                        }
+
+                        $feature = $dependency->getFeature();
+                        if ($feature instanceof UnresolvedFeature) {
+                            $feature = $features[$feature->getName()] ?? null;
+                            if ($feature === null) {
+                                $dependencies[] = $dependency;
+                                continue;
+                            }
+                        }
+
+                        $requiredVariants = [];
+                        foreach ($dependency->getRequiredVariants() ?? [] as $requiredVariant) {
+                            if (!$requiredVariant instanceof UnresolvedVariant) {
+                                $requiredVariants[] = $requiredVariant;
+                                continue;
+                            }
+
+                            foreach ($feature->getVariants() as $variant) {
+                                if ($variant->getName() !== $requiredVariant->getName()) {
+                                    continue;
+                                }
+                                $requiredVariant = $variant;
+                                break;
+                            }
+                            $requiredVariants[] = $requiredVariant;
+                        }
+
+                        $dependencies[] = new DefaultFeatureDependency(
+                            feature: $feature,
+                            expectedState: $dependency->getExpectedState(),
+                            requiredVariants: $requiredVariants,
+                        );
+                    }
+
+                    $features[$unresolvedItem->getName()] = new DefaultFeature(
+                        name: $unresolvedItem->getName(),
+                        enabled: $unresolvedItem->isEnabled(),
+                        strategies: $unresolvedItem->getStrategies(),
+                        variants: $unresolvedItem->getVariants(),
+                        impressionData: $unresolvedItem->hasImpressionData(),
+                        dependencies: $dependencies,
+                    );
+                    unset($unresolved[$index]);
+                }
+            } while (count($unresolved) < $currentUnresolvedCount);
+        }
+    }
+
+    /**
+     * @param array<array{feature: string, enabled?: bool, variants?: array<string>}> $dependencies
+     * @param array<string, Feature>                                                  $features
+     *
+     * @return array<FeatureDependency>
+     */
+    private function parseDependencies(array $dependencies, array $features, ?bool &$hasUnresolvedDependencies = null): array
```

Comment:
and same here, personally I'd do extract method refactoring to explain the steps of the algorithm and reduce the indentation level/whitespace on the left hand side of the code.

Reply:
I tried to split it up a bit, hopefully it helps.